src/currencyservice/src/server.cpp

```diff
@@ -0,0 +1,250 @@
+#include <iostream>
+#include <math.h>
+#include <demo.grpc.pb.h>
+#include <grpc/health/v1/health.grpc.pb.h>
+
+#include "opentelemetry/trace/context.h"
+#include "opentelemetry/trace/experimental_semantic_conventions.h"
+#include "opentelemetry/trace/span_context_kv_iterable_view.h"
+#include "tracer_common.h"
+
+#include <grpcpp/grpcpp.h>
+#include <grpcpp/server.h>
+#include <grpcpp/server_builder.h>
+#include <grpcpp/server_context.h>
+
+using namespace std;
+
+using hipstershop::Empty;
+using hipstershop::GetSupportedCurrenciesResponse;
+using hipstershop::CurrencyConversionRequest;
+using hipstershop::Money;
+
+using grpc::Status;
+using grpc::ServerContext;
+using grpc::ServerBuilder;
+using grpc::Server;
+
+using Span        = opentelemetry::trace::Span;
+using SpanContext = opentelemetry::trace::SpanContext;
+using namespace opentelemetry::trace;
+
+namespace context = opentelemetry::context;
+
+namespace
+{
+  std::unordered_map<std::string, double> currency_conversion{
+    {"EUR", 1.0},
+    {"USD", 1.1305},
+    {"JPY", 126.40},
+    {"BGN", 1.9558},
+    {"CZK", 25.592},
+    {"DKK", 7.4609},
+    {"GBP", 0.85970},
+    {"HUF", 315.51},
+    {"PLN", 4.2996},
+    {"RON", 4.7463},
+    {"SEK", 10.5375},
+    {"CHF", 1.1360},
+    {"ISK", 136.80},
+    {"NOK", 9.8040},
+    {"HRK", 7.4210},
+    {"RUB", 74.4208},
+    {"TRY", 6.1247},
+    {"AUD", 1.6072},
+    {"BRL", 4.2682},
+    {"CAD", 1.5128},
+    {"CNY", 7.5857},
+    {"HKD", 8.8743},
+    {"IDR", 15999.40},
+    {"ILS", 4.0875},
+    {"INR", 79.4320},
+    {"KRW", 1275.05},
+    {"MXN", 21.7999},
+    {"MYR", 4.6289},
+    {"NZD", 1.6679},
+    {"PHP", 59.083},
+    {"SGD", 1.5349},
+    {"THB", 36.012},
+    {"ZAR", 16.0583},
+};
+
+class HealthServer final : public grpc::health::v1::Health::Service
+{
+  Status Check(
+    ServerContext* context,
+    const grpc::health::v1::HealthCheckRequest* request,
+    grpc::health::v1::HealthCheckResponse* response) override
+  {
+    response->set_status(grpc::health::v1::HealthCheckResponse::SERVING);
+    return Status::OK;
+  }
+};
+
+class CurrencyService final : public hipstershop::CurrencyService::Service
+{
+  Status GetSupportedCurrencies(ServerContext* context,
+  	const Empty* request,
+  	GetSupportedCurrenciesResponse* response) override
+  {
+    StartSpanOptions options;
+    options.kind = SpanKind::kServer;
+    GrpcServerCarrier carrier(context);
+
+    auto prop        = context::propagation::GlobalTextMapPropagator::GetGlobalPropagator();
+    auto current_ctx = context::RuntimeContext::GetCurrent();
+    auto new_context = prop->Extract(carrier, current_ctx);
+    options.parent   = GetSpan(new_context)->GetContext();
+
+    std::string span_name = "CurrencyService/GetSupportedCurrencies";
+    auto span =
+        get_tracer("opentelemetry-cpp")->StartSpan(span_name,
```

Comment:
I believe we've been naming tracer's with the service names in other services. so this would be `"currencyservice"`